File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        TreeNode prev = null;
        while(root != null) {
            if(root.left == null) {
                if(prev != null && prev.val >= root.val) {
                    return false;
                }
                prev = root;
                root = root.right;
            }
            else {
                TreeNode curr = root.left;
                while(curr.right != null && curr.right != root) {
                    curr = curr.right;
                }
                if(curr.right == null) {
                    curr.right = root;
                    root = root.left;
                }
                else {
                    curr.right = null;
                    if(prev != null && prev.val >= root.val) {
                        return false;
                    }
                    prev = root;
                    root = root.right;
                }
            }
        }
        return true;
    }
}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getPredecessor(self, node: Optional[TreeNode]) -> Optional[TreeNode]:
        curr = node.left
        while curr.right and curr.right != node:
            curr = curr.right
        return curr

    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        prev = float('-inf')
        curr = root
        while curr:
            if curr.left:
                predecessor = self.getPredecessor(curr)
                if predecessor.right:
                    predecessor.right = None
                    if prev >= curr.val:
                        return False
                    prev = curr.val
                    curr = curr.right
                else:
                    predecessor.right = curr
                    curr = curr.left
            else:
                if prev >= curr.val:
                    return False
                prev = curr.val
                curr = curr.right
        return True